vivisect/analysis/amd64/emulation.py

```diff
@@ -33,96 +31,62 @@ def prehook(self, emu, op, starteip):
             if len(op.opers):
                 self.retbytes = op.opers[0].imm
 
-##### FIXME: this should be all done through the calling convention
-sysvamd64argnames = {
-    0: ('rdi', e_amd64.REG_RDI),
-    1: ('rsi', e_amd64.REG_RSI),
-    2: ('rdx', e_amd64.REG_RDX),
-    3: ('rcx', e_amd64.REG_RCX),
-    4: ('r8',  e_amd64.REG_R8),
-    5: ('r9',  e_amd64.REG_R9),
-}
-
-msx64argnames = {
-    0: ('rcx', e_amd64.REG_RCX),
-    1: ('rdx', e_amd64.REG_RDX),
-    2: ('r8',  e_amd64.REG_R8),
-    3: ('r9',  e_amd64.REG_R9),
-}
-
-arch_bindings = {
-    'msx64call': msx64argnames,
-    'sysvamd64call': sysvamd64argnames,
-    None: [],
-}
-
-def sysvamd64name(idx):
-    ret = sysvamd64argnames.get(idx)
-    if ret == None:
-        name = 'arg%d' % idx
-    else:
-        name, idx = ret
-    return name
-
-def msx64name(idx):
-    ret = msx64argnames.get(idx)
-    if ret == None:
-        name = 'arg%d' % idx
-    else:
-        name, idx = ret
-    return name
-#####
-
 def buildFunctionApi(vw, fva, emu, emumon):
-    
+    '''
+    Builds the function API:
+    * argc - number of arguments
+    * cc   - determines calling convention
+    '''
     argc = 0
     funcargs = []
     callconv = vw.getMeta('DefaultCall')
-    argnames = arch_bindings.get(callconv)
+    cc = emu.getCallingConvention(callconv)
+    argnames = cc.getCallRegArgInfo(emu)
     undefregs = set(emu.getUninitRegUse())
 
+    ### determine argument count, register and stack
+    # determine number of register args
     for argnum in range(len(argnames), 0, -1):
-        argname, argid = argnames[argnum-1]
+        #argname, argid = argnames[argnum-1]
+        argtype, argname, argid = argnames[argnum-1]
         if argid in undefregs:
             argc = argnum
             break
 
+    # calculate the stack argument space used
+    if emumon.stackmax > cc.pad:
+        # stack delta - padding (shadow space, etc)
+        stackargspace = emumon.stackmax - cc.pad
+        targc = cc.getNumRegArgs(emu, MAX_ARGS) + (stackargspace / 8)
+
+        if targc > MAX_ARGS:
+            emumon.logAnomaly(emu, fva, 'Crazy Stack Offset Touched: 0x%.8x' % emumon.stackmax)
+        else:
+            argc = targc
+
+    # add in shadow space for msx64call
```

Comment:
all of the above code abstracts the CC-specific code and puts the details into the CallingConvention object.